### PR TITLE
docs: add further snippets with new infrastructure

### LINKED ISSUE #2947: Migrate snippets from `guide/create-fuels`
- [X] `decrement_counter.test.ts`

### LINKED ISSUE #2948: Migrate snippets from `guide/encoding`
- [X] `encode-and-decode.test.ts`
- [X] `working-with-bytes.test.ts`

### LINKED ISSUE #2952: Migrate snippets from `guide/scripts`
- [X] `script-custom-transaction.test.ts`
- [X] `script-with-configurable.test.ts`

### LINKED ISSUE #2953: Migrate snippets from `guide/testing`
- [ ] `launching-a-test-node.test.ts`
- [ ] `tweaking-the-blockchain.test.ts`

### LINKED ISSUE #2956: Migrate snippets from `guide/utilities`
- [X] `unit-conversion.test.ts`
- [X] `getting-started-with-wallet-manager.test.ts`
- [X] `locking-and-unlocking-wallet-manager.test.ts`

### LINKED ISSUE #2957: Migrate snippets from `guide/wallet-manager`
- [X] getting-started-with-wallet-manager.test.ts
- [X] locking-and-unlocking-wallet-manager.test.ts

### PR DESCRIPTION
**Pull request by maschad**
- Related to #2399
 - Related to #2503
 - Closes #2948
 - Closes #2947
 - Closes #2957
 - Closes #2956
 - Closes #2953
 - Closes #2952

# Summary

This PR builds on the work @arboleya did in https://github.com/FuelLabs/fuels-ts/pull/2651 by introducing some more snippets that have been migrated to the new infra. Once this is merged into that branch then the docs can be refactored to use the folder structure in https://github.com/FuelLabs/fuels-ts/issues/3205

# Checklist

- [x] All **changes** are **covered** by **tests** (or not applicable)
- [x] All **changes** are **documented** (or not applicable)
- [x] I **reviewed** the **entire PR** myself (preferably, on GH UI)
- [x] I **described** all **Breaking Changes** (or there's none)

### CODE REVIEW

File: apps/docs-snippets2/scripts/test-template.ts [hunk 1]
```diff
@@ -10,7 +10,6 @@ afterAll(() => {
 
 /**
  * @group node
- * @group browser
```
Comment (maschad) [inline apps/docs-snippets2/scripts/test-template.ts:13]: I removed browser testing on these snippets but perhaps we just need to add a configuration to each snippet to determine whether it's node or browser specific, or compatible in both environments

File: apps/docs-snippets2/src/create-fuels/decrement-counter.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+// Lets setup some values to use in the test.
+const initialCount = 0;
+const incrementedValue = 5;
+const decrementedValue = 2;
+
+// We can now call the contract functions and test the results. Lets assert the initial value of the counter.
+const { waitForResult: initWaitForResult } = await contract.functions.get_count().call();
+const { value: initValue } = await initWaitForResult();
+console.assert(
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/create-fuels/decrement-counter.ts:24]: @maschad didn't know about this method, nice!

File: apps/docs-snippets2/src/create-fuels/decrement-counter.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+const { contract } = await waitForResult();
+
+// Lets setup some values to use in the test.
+const initialCount = 0;
+const incrementedValue = 5;
+const decrementedValue = 2;
+
+// We can now call the contract functions and test the results. Lets assert the initial value of the counter.
+const { waitForResult: initWaitForResult } = await contract.functions.get_count().call();
+const { value: initValue } = await initWaitForResult();
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/create-fuels/decrement-counter.ts:23]:
```suggestion
const { waitForResult: getCountWaitForResult } = await contract.functions.get_count().call();
const { value: initialGetCountValue } = await getCountWaitForResult();
```

Nit

File: apps/docs-snippets2/src/create-fuels/decrement-counter.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+console.assert(
+  initValue.toNumber() === initialCount,
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/create-fuels/decrement-counter.ts:25]:
```suggestion
  initialGetCountValue.toNumber() === initialCount,
```

File: apps/docs-snippets2/src/encoding/working-with-bytes.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  'Encoded U64 should be equal to 255'
+);
+console.assert(
+  encodedU256.toString() ===
+    new Uint8Array([
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      255,
+    ]).toString(),
+  'Encoded U256 should be equal to 255'
+);
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/encoding/working-with-bytes.ts:67]: Is it worth moving these asserts alongside the calls, then we can remove the comments.

Example:
```ts
const u8Coder = new NumberCoder('u8');
const encodedU8 = u8Coder.encode(255);

console.assert(
  encodedU8.toString() === new Uint8Array([255]).toString(),
  'Encoded U8 should be equal to 255'
);
```

File: apps/docs-snippets2/src/utilities/unit-conversion.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import { BN, DECIMAL_GWEI, DECIMAL_KWEI, bn, Provider, Wallet } from 'fuels';
+
+import { LOCAL_NETWORK_URL, WALLET_PVT_KEY } from '../env';
+import { EchoValuesFactory } from '../typegend/contracts/EchoValuesFactory';
+
+// #region instantiation-1
+const myBigNumberOne = '100000000';
+
+const resultOne = new BN('100000000').toString();
+console.assert(resultOne === myBigNumberOne, 'BN is not equal to expected');
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/utilities/unit-conversion.ts:11]:
```suggestion
const resultOne = new BN('100000000').toString();

console.assert(resultOne === myBigNumberOne, 'BN is not equal to expected');
```

Maybe some whitespace between these will help readability. Same goes for all on this file.

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+});
+
+const { provider: providerWithTestMessagesOnChain } = launchedWithTestMessagesOnChain;
+
+recipient.provider = providerWithTestMessagesOnChain;
+
+const {
+  messages: [messageOnChain],
+} = await recipient.getMessages();
+// message.nonce === testMessage.nonce
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:219]:
```suggestion

```

Personal don't think the comment is required now

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+console.assert(
+  coins[0].assetId === randomAssetIds[0].value,
+  'Asset id is not randomAssetIds[0].value'
+);
+// #endregion asset-ids
+
+nodeWithCustomAssetIds.cleanup();
+
+// #region test-messages
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:170]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+      },
+    },
+  },
+});
+// #endregion custom-node-options
+
+nodeWithCustomBaseAssetId.cleanup();
+
+// #region asset-ids
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:147]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+const { provider: providerWithCustomChainConfig } = launchedWithCustomChainConfig;
+
+const { name } = await providerWithCustomChainConfig.fetchChain();
+
+console.assert(name === 'custom-chain-config', 'Chain name is not custom-chain-config');
+// #endregion custom-chain-config
+
+launchedWithCustomChainConfig.cleanup();
+// #region custom-node-options
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:126]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+);
+console.assert(wallet1 instanceof WalletUnlocked, 'Wallet1 is not instance of WalletUnlocked');
+console.assert(wallet2 instanceof WalletUnlocked, 'Wallet2 is not instance of WalletUnlocked');
+console.assert(wallet3 instanceof WalletUnlocked, 'Wallet3 is not instance of WalletUnlocked');
+console.assert(wallet4 instanceof WalletUnlocked, 'Wallet4 is not instance of WalletUnlocked');
+
+// #endregion advanced-example
+
+// #region custom-fuel-core-args
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:96]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+const { waitForResult } = await contract.functions.get_count().call();
+const response = await waitForResult();
+
+console.assert(response.value.toNumber() === 0, 'Counter is not initialized');
+console.assert(provider instanceof Provider, 'Provider is not instance of Provider');
+console.assert(wallets.length === 4, 'Wallets length is not 4');
+// #endregion basic-example
+
+// #region advanced-example
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:58]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+/*
+      Do your things, run your tests, and then call
+      `launchedTestNode.cleanup()` to dispose of everything.
+    */
+
+launchedTestNode.cleanup();
+// #endregion manual-cleanup
+
+// #region options
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:31]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+/*
+ * The method `launched.cleanup()` will be automatically
+ * called when the variable `launched` goes out of block scope.
+ */
+
+// #endregion automatic-cleanup
+
+// #region manual-cleanup
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:19]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts [hunk 15]
```diff
@@ -0,0 +1,224 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable import/first */
+import { WalletUnlocked, Provider } from 'fuels';
+import { launchTestNode, TestAssetId, TestMessage } from 'fuels/test-utils';
+import { join } from 'path';
+
+// #region automatic-cleanup
+
```
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:8]:
```suggestion
```

File: apps/docs-snippets2/src/testing/launching-a-test-node.node.ts
Comment (petertonysmith94) [inline apps/docs-snippets2/src/testing/launching-a-test-node.node.ts:1]: Sorry for all the whitespace nits here.

**Review by petertonysmith94 — APPROVED** [review]
@maschad this is absolute 🔥 Big kudos!

I've left a couple nits, but please feel free to ignore as overall, I'm super happy with this (approving on this basis).

One question I did have, was why we're using `new Script` instantiation over typegen?

File: apps/docs-snippets2/src/encoding/working-with-bytes.ts (hunk 5 above)
Comment (maschad) [reply apps/docs-snippets2/src/encoding/working-with-bytes.ts]: 7f38a0cb5

**Review by petertonysmith94 — APPROVED** [review]

File: apps/docs-snippets2/src/encoding/working-with-bytes.ts [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+const u256Coder = new BigNumberCoder('u256');
+const encodedU256 = u256Coder.encode(255);
+console.assert(
+  encodedU256.toString() ===
+    new Uint8Array([
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      255,
+    ]).toString(),
```
Comment (Torres-ssf) [inline apps/docs-snippets2/src/encoding/working-with-bytes.ts:57]: @maschad What is the motivation behind using `console.asset` here?

This approach has increased the verbosity of some code snippets.

Before:
Screenshot 2024-10-02 at 10 05 57

After:
Screenshot 2024-10-02 at 10 06 31

**Review by Torres-ssf — COMMENTED** [review]
Great job on this 👏 🚀

File: apps/docs-snippets2/src/encoding/working-with-bytes.ts (hunk 16 above)
Comment (arboleya) [reply apps/docs-snippets2/src/encoding/working-with-bytes.ts:57]: In fact, we should not use `console.assert` or `expect` ([here](https://github.com/FuelLabs/fuels-ts/pull/3233/files#diff-2ade8bacc0cbacf42b1a485e15be9fd666c0f0ec0c3f57ce64dbc2026547604eR57-R58)) in the snippets anymore.

Only `console.log` and those will be converted to test assertions.

Let's do another quick sync over this? @maschad @Torres-ssf

Comment (Torres-ssf) [reply apps/docs-snippets2/src/encoding/working-with-bytes.ts:57]: @arboleya Sure!

While working on [FuelLabs/fuels-ts#3233](https://github.com/FuelLabs/fuels-ts/pull/3233), I aimed to keep the code snippets as they are. However, I'm happy to remove the test `expects` if that’s the preferred direction.

Comment (arboleya) [reply apps/docs-snippets2/src/encoding/working-with-bytes.ts:57]: We also had this issue to remove assertions from the snippets:
 - https://github.com/FuelLabs/fuels-ts/issues/1138

I [closed it](https://github.com/FuelLabs/fuels-ts/issues/1138#issuecomment-2373700450), but the rationale remains for the new snippets.

File: apps/docs-snippets2/scripts/wrap-snippets.ts [hunk 17]
```diff
@@ -12,6 +12,20 @@ const wrapperFnContents = readFileSync(wrapperFnFilepath, 'utf-8');
...
+
+  // Check if the filepath includes '.node' or '.browser'
+  if (filepath.includes('.node')) {
+    testEnvironments = '/**\n * @group node\n */';
+  } else if (filepath.includes('.browser')) {
+    testEnvironments = '/**\n * @group browser\n */';
+  } else {
+    testEnvironments = '/**\n * @group node\n * @group browser\n */';
+  }
+
```
Comment (arboleya) [inline apps/docs-snippets2/scripts/wrap-snippets.ts:28]: We should start documenting these things in a `docs-snippets2/README.md` file.

File: tsconfig.test.json [hunk 18]
```diff
@@ -4,5 +4,5 @@
     "noEmit": true
   },
   "include": ["**/*.test.ts"],
-  "exclude": ["node_modules"]
+  "exclude": ["node_modules", "apps/docs-snippets2/src/**/*.test.ts"]
```
Comment (arboleya) [inline tsconfig.test.json:7]: Why was this necessary?

File: .eslintignore [hunk 19]
```diff
@@ -13,6 +13,7 @@ apps/demo-react-cra
 apps/demo-react-vite
 apps/create-fuels-counter-guide
 apps/docs-snippets2/src/typegend
+apps/docs-snippets2/src/**/*.test.ts
```
Comment (arboleya) [inline .eslintignore:16]: Is the same required for prettier?

File: apps/docs-snippets2/scripts/test-template.ts [hunk 20]
```diff
@@ -8,10 +8,7 @@ afterAll(() => {
   console = { ...consoleBkp };
 });
 
-/**
- * @group node
- * @group browser
- */
+// %TEST_ENVIRONMENT%
```
Comment (arboleya) [inline apps/docs-snippets2/scripts/test-template.ts:11]: @maschad Regarding the PR description and according to [#3157](https://github.com/FuelLabs/fuels-ts/issues/3157)'s premises:

> GitHub issues must be informed one per line (predictable parsing)

Regexes [here](https://github.com/FuelLabs/github-actions/blob/9de966c6488468852bd7418420c781fc1f3e91da/linear/link-pr-to-linear-issue/src/link-pr-to-linear-issue.ts#L10) and [here](https://github.com/FuelLabs/github-actions/blob/9de966c6488468852bd7418420c781fc1f3e91da/linear/link-pr-to-linear-issue/src/link-pr-to-linear-issue.ts#L16).

Comment (arboleya) [reply apps/docs-snippets2/scripts/test-template.ts:11]: @nedsalk Should we add logs and basic validations that fail the action so we know something is wrong and needs to be corrected?

Comment (maschad) [reply apps/docs-snippets2/scripts/test-template.ts:11]: @arboleya thanks for highlighting that (as an aside I don't think your comment relates to line 11 of this file), I've updated the PR description.

But it seems that Linear had picked it up regardless 🤷🏾

File: tsconfig.test.json (hunk 18 above)
Comment (maschad) [reply tsconfig.test.json:7]: Strangely, this was the only way I could get the linter to ignore top-level _await_ despite us [enabling ES2022](https://github.com/FuelLabs/fuels-ts/blob/mc/continue-infra-for-snippets/apps/docs-snippets2/tsconfig.json#L4) for this package, not sure why though.

File: .eslintignore (hunk 19 above)
Comment (maschad) [reply .eslintignore:16]: That was already added in @Torres-ssf 's [PR](https://github.com/FuelLabs/fuels-ts/pull/3238/files#diff-663ade211b3a1552162de21c4031fcd16be99407aae5ceecbb491a2efc43d5d2R30)

File: apps/docs-snippets2/scripts/wrap-snippets.ts (hunk 17 above)
Comment (maschad) [reply apps/docs-snippets2/scripts/wrap-snippets.ts:28]: e19b126ef

File: apps/docs-snippets2/scripts/test-template.ts (hunk 20 above)
Comment (nedsalk) [reply apps/docs-snippets2/scripts/test-template.ts:11]: > @arboleya 
> @nedsalk Should we add logs and basic validations that fail the action so we know something is wrong and needs to be corrected?

I'm not sure that it's worthwhile to add these checks. I think it's enough that we have it explained in the [readme of the action](https://github.com/FuelLabs/github-actions/tree/master/linear/link-pr-to-linear-issue) and the users should acquaint themselves with it. GitHub also doesn't link issues to PRs when multiple are listed in one line (e.g. `Closes #abc, #def`) and it's up to the developer to know that; once they fix it to link properly in GitHub, the action will pick it up as well.

@maschad You should update the description for the related issues and change it from `Related` to `Related to / Relates to / Part of` because those are the only magic words that our action picks up.

> But it seems that Linear had picked it up regardless 🤷🏾

I just took a look in Linear and this PR isn't linked to the issues. What's weird is that our linear workflow didn't even run on this branch. :thinking:

Comment (arboleya) [reply apps/docs-snippets2/scripts/test-template.ts:11]: All these explanations can be automatic validations.

Comment (nedsalk) [reply apps/docs-snippets2/scripts/test-template.ts:11]: @arboleya sure, I just thought that it might be overkill. If you feel that it should be done, please create the issue and we can pick it up.

File: tsconfig.test.json (hunk 18 above)
Comment (arboleya) [reply tsconfig.test.json:7]: Hmm it looks like some misconfig.

The problem with ignoring these files is that it'll silence everything.

Comment (maschad) [reply tsconfig.test.json:7]: This file is responsible for compiling the test files and so it overrides any subdirectory _tsconfigs_ , including the one local to `docs-snippets2` 

But given we already run the compiler against the snippet files themselves, we shouldn't really need to run the typescript compiler against the generated test files, since they aren't tracked in the repo and any potential issues would result in test failures regardless.

**Comment by maschad** [comment]
Closed in favour of https://github.com/FuelLabs/fuels-ts/pull/3297
